### PR TITLE
[WIP] File IO with UI

### PR DESCRIPTION
**Pull request by paul1956**
Support for VB FileIO with UI on Windows and exception on other OS's

### CODE REVIEW

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/FileIO/FileSystem.Windows.vb [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        ''' Copy/MoveFile will call this directly. Copy/MoveDirectory will call ShellCopyOrMoveDirectory first
+        ''' to change the path if needed.
+        ''' </remarks>
+        Private Shared Sub ShellCopyOrMove(ByVal Operation As CopyOrMove, ByVal TargetType As FileOrDirectory,
+            ByVal FullSourcePath As String, ByVal FullTargetPath As String, ByVal ShowUI As UIOptionInternal, ByVal OnUserCancel As UICancelOption)
+            Debug.Assert(System.Enum.IsDefined(GetType(CopyOrMove), Operation))
+            Debug.Assert(System.Enum.IsDefined(GetType(FileOrDirectory), TargetType))
+            Debug.Assert(FullSourcePath.Length <> 0 AndAlso IO.Path.IsPathRooted(FullSourcePath), "Invalid FullSourcePath!!!")
+            Debug.Assert(FullTargetPath.Length <> 0 AndAlso IO.Path.IsPathRooted(FullTargetPath), "Invalid FullTargetPath!!!")
+            Debug.Assert(ShowUI <> UIOptionInternal.NoUI, "Why call ShellDelete if ShowUI is NoUI???")
```
Comment (Wraith2) [inline src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/FileIO/FileSystem.Windows.vb]: Doesn't ShellDelete allow the deleted file to be put into the recycle bin where a direct delete operation won't?

Comment (paul1956) [reply src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/FileIO/FileSystem.Windows.vb]: @Wraith2 The code in this file and all the src files is original, I am just writing tests and making minimal changes to be consistent with CoreFx and to split for OS specific issues. The Assert is in the original code.

Comment (paul1956) [reply src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/FileIO/FileSystem.Windows.vb]: @Wraith2 actually some of the Asserts are wrong in the original source but I guess were never tested. TargetPath is tested for Length<>0 but it is valid to be Nothing in the case of a Delete operation.

**Comment by paul1956** [comment]
@danmosemsft @jaredpar This is at a point that it could be reviewed. I have manual and automatic tests for all the FileIO functions. I also added MsgBox support and made a few changes in existing code to support the changes I needed to get MsgBox to work, those changes needs extra review and are located in Interaction.vb. There are other functions in FileIO like TextParsing and I have not done anything with that except getting it to compile. I would not bother reviewing. A lot of functions are straight passthroughs to System.IO and I am not sure how much additional testing is needed or how to let reviewers know that fact.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/FileIO/FileSystem.Windows.vb [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        ''' Calls NativeMethods.SHFileOperation with the given SHFILEOPSTRUCT, notifies the shell of change,
+        ''' and throw exceptions if needed.
+        ''' </summary>
+        ''' <param name="OperationType">Value from SHFileOperationType, specifying Copy / Move / Delete</param>
+        ''' <param name="OperationFlags">Value from ShFileOperationFlags, specifying overwrite, recycle bin, etc...</param>
+        ''' <param name="FullSource">The full path to the source.</param>
+        ''' <param name="FullTarget">The full path to the target. Nothing if this is a Delete operation.</param>
+        ''' <param name="OnUserCancel">Value from UICancelOption, specifying to throw or not when user cancels the operation.</param>
+        ''' <remarks>
+        ''' !!!!! SECURITY WARNING !!!!
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/FileIO/FileSystem.Windows.vb]: Everything to do with System.Security.Permissions can be removed including this comment and any strings constructed or work performed only for that purpose such as CheckPermissionPath below. There may be others, I didn't look.

File: src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Interaction.Windows.vb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+            Return CType(NativeMethods.MessageBox(ParentWindow, sPrompt, sTitle, CUInt(Buttons)), MsgBoxResult)
+        End Function
+
+        Private Shared Function GetTitleFromAssembly(ByVal CallingAssembly As Reflection.Assembly) As String
+
+            Dim Title As String
+
+            'Get the Assembly name of the calling assembly 
+            'Assembly.GetName requires PathDiscovery permission so we try this first
```
Comment (danmosemsft) [inline src/Microsoft.VisualBasic/src/Microsoft/VisualBasic/Interaction.Windows.vb]: More CAS security related comments to remove

**Comment by paul1956** [comment]
@333fred I am not sure this is worth reviewing because for now it will not build since  
"Merge branch 'master' into FileIO" and I don't know how to undo it. I would love some help getting the build to work if anyone is available. I did the merge to resolve a conflict in the ref file and was causing the build to fail.

**Comment by danmosemsft** [comment]
```
Microsoft.VisualBasic.cs(116,6): error CS0246: The type or namespace name 'FlagsAttribute' could not be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---74aa877a\src\Microsoft.VisualBasic\ref\Microsoft.VisualBasic.csproj]
Microsoft.VisualBasic.cs(116,6): error CS0246: The type or namespace name 'Flags' could not be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---74aa877a\src\Microsoft.VisualBasic\ref\Microsoft.VisualBasic.csproj]
Microsoft.VisualBasic.cs(960,44): error CS0246: The type or namespace name 'IDisposable' could not be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---74aa877a\src\Microsoft.VisualBasic\ref\Microsoft.VisualBasic.csproj]
Microsoft.VisualBasic.cs(1051,14): error CS0246: The type or namespace name 'IDisposable' could not be found (are you missing a using directive or an assembly reference?) [D:\j\workspace\windows-TGrou---74aa877a\src\Microsoft.VisualBasic\ref\Microsoft.VisualBasic.csproj]
Microsoft.VisualBasic.cs(1051,14): error CS0538: 'IDisposable' in explicit interface declaration is not an interface [D:\j\workspace\windows-TGrou---74aa877a\src\Microsoft.VisualBasic\ref\Microsoft.VisualBasic.csproj]
```

**Comment by danmosemsft** [comment]
There is no using statement (Imports) in ref files so everything must be fully qualified eg `System.IDisposable`
Also the file should be formatted consistently like the other ref files as sometimes it is regenerated by a script. Eg braces on the same line etc

**Comment by paul1956** [comment]
I am closing this PR, I will open another one starting with latest master

**Comment by danmosemsft** [comment]
@paul1956 if you can, please try not to make duplicate PRs. A rebase of some sort and a force push should always be sufficient to get a PR back in a good state. The main reason to open a new PR that I can think of might be because conversation has gotten toxic, off topic or much too long.

Rebasing and merging can take some practice before it becomes familiar (it did for me) but bear in mind that you can "xcopy" your local repo to a scratch location where you safely can try different commands to see what they do.

**Comment by paul1956** [comment]
@danmosemsft I tried all those things but things were so corrupt I could not get it to build and could not do a commit. I am trying to be much more careful so hopefully I will not do it again. I am not sure I understand your xcopy comment. Are you saying that if I make a copy I can do local Git commands that will not effect the image on GitHub?

**Comment by paul1956** [comment]
Should I delete the 2 closed branches, all the feedback is in new PR.

**Comment by danmosemsft** [comment]
Every clone of the repo, including yours, is essentially equivalent - everything can be reconstructed from it. You can create a copy of your local clone by using `git clone` but for the purposes of local experimentation it should be possible to just `xcopy` the entire clone (the folder you likely named corefx) into another folder. 

(At this point you can safely do almost anything -- the only git command that I'm aware of that affects the server is "git push", and even then, in the context of github, that can only affect your fork and not the main repo because you don't have permissions to push to it.)

Doing this will give you a playground where you can play with merge, rebase etc and see what results you get using various permutations of git log, etc. I do recommend you spend some time experimenting -- and you may find it helpful to browse some walkthroughs -- until you are able to use rebase and merge fairly fluently. If you mess up your playground - just delete it and create another.

You may find a graphical visualization tool helpful - I have used SourceTree in the past, or gitk.

As far as deleting the closed branches, that would be tidy, but it does not matter to me as they are in your fork, not the main repo. PR's however cannot  be deleted, only closed, which is why it's best to not create them unnecessarily.

**Pull request closed without merging**